File: app/src/main/java/com/byd/dashcast/proxy/daemon/FissionWatchdogRegistry.java
package com.byd.dashcast.proxy.daemon;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

/** Tracks the latest watchdog generation independently for each Layout package. */
final class FissionWatchdogRegistry {
    private final AtomicLong sequence = new AtomicLong();
    private final ConcurrentHashMap<String, Long> generations = new ConcurrentHashMap<>();

    long start(String packageName) {
        if (packageName == null || packageName.isEmpty()) {
            throw new IllegalArgumentException("packageName required");
        }
        long generation = sequence.incrementAndGet();
        generations.put(packageName, generation);
        return generation;
    }

    boolean isCurrent(String packageName, long generation) {
        Long current = generations.get(packageName);
        return current != null && current == generation;
    }

    void finish(String packageName, long generation) {
        generations.remove(packageName, generation);
    }

    boolean cancel(String packageName) {
        return packageName != null && generations.remove(packageName) != null;
    }

    int cancelAll() {
        int count = generations.size();
        generations.clear();
        return count;
    }

    int activePackages() {
        return generations.size();
    }
}

File: app/src/main/java/com/byd/dashcast/proxy/daemon/FissionWatchdogRegistry.kt
package com.byd.dashcast.proxy.daemon

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/** Tracks the latest watchdog generation independently for each Layout package. */
internal class FissionWatchdogRegistry {
    private val sequence = AtomicLong()
    private val generations = ConcurrentHashMap<String, Long>()

    fun start(packageName: String?): Long {
        require(!packageName.isNullOrEmpty()) { "packageName required" }
        val generation = sequence.incrementAndGet()
        generations[packageName] = generation
        return generation
    }

    fun isCurrent(packageName: String?, generation: Long): Boolean {
        val current = packageName?.let { generations[it] }
        return current != null && current == generation
    }

    fun finish(packageName: String?, generation: Long) {
        if (packageName != null) generations.remove(packageName, generation)
    }

    fun cancel(packageName: String?): Boolean {
        return packageName != null && generations.remove(packageName) != null
    }

    fun cancelAll(): Int {
        val count = generations.size
        generations.clear()
        return count
    }

    fun activePackages(): Int = generations.size
}
